GitCliUtil.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"fmt"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+type GitUtil struct {
+}
+
+func NewGitUtil() *GitUtil {
+	return &GitUtil{
+	}
+}
+
+const GIT_AKS_PASS = "/git-ask-pass.sh"
+
+func (impl *GitUtil) Fetch(rootDir string, username string, password string) (response, errMsg string, err error) {
+	log.Println(devtron, "git fetch ", "location", rootDir)
+	cmd := exec.Command("git", "-C", rootDir, "fetch", "origin", "--tags", "--force")
+	output, errMsg, err := impl.runCommandWithCred(cmd, username, password)
+	log.Println(devtron, "fetch output", "root", rootDir, "opt", output, "errMsg", errMsg, "error", err)
+	return output, "", nil
+}
+
+func (impl *GitUtil) Checkout(rootDir string, branch string) (response, errMsg string, err error) {
+	log.Println(devtron, "git checkout ", "location", rootDir)
+	cmd := exec.Command("git", "-C", rootDir, "checkout", branch, "--force")
+	output, errMsg, err := impl.runCommand(cmd)
+	log.Println(devtron, "checkout output", "root", rootDir, "opt", output, "errMsg", errMsg, "error", err)
+	return output, "", nil
+}
+
+func (impl *GitUtil) runCommandWithCred(cmd *exec.Cmd, userName, password string) (response, errMsg string, err error) {
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("GIT_ASKPASS=%s", GIT_AKS_PASS),
+		fmt.Sprintf("GIT_USERNAME=%s", userName), // ignored
+		fmt.Sprintf("GIT_PASSWORD=%s", password), // this value is used
+	)
+	return impl.runCommand(cmd)
+}
+
+func (impl *GitUtil) runCommand(cmd *exec.Cmd) (response, errMsg string, err error) {
+	cmd.Env = append(cmd.Env, "HOME=/dev/null")
+	outBytes, err := cmd.CombinedOutput()
+	if err != nil {
+		exErr, ok := err.(*exec.ExitError)
+		if !ok {
+			return "", "", err
+		}
+		errOutput := string(exErr.Stderr)
+		return "", errOutput, err
+	}
+	output := string(outBytes)
+	output = strings.Replace(output, "\n", "", -1)
+	output = strings.TrimSpace(output)
+	return output, "", nil
+}
+
+func (impl *GitUtil) Init(rootDir string, remoteUrl string, isBare bool) error {
+
+	//-----------------
+
+	err := os.MkdirAll(rootDir, 0755)
```

Comment:
if called repeatedly will give  error

Reply:
will be called once per repo, so should not be a problem